Sources/Purchasing/Purchases/PurchaseParams.swift

```diff
@@ -0,0 +1,72 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  PurchaseParams.swift
+//
+//  Created by mark on 17/10/24.
+
+import Foundation
+
+#if ENABLE_PURCHASE_PARAMS
+
+/**
+ * ``PurchaseParams`` can be used to add configuration options when making a purchase.
+ * This class follows the builder pattern.
+ *
+ * Example making a purchase using ``PurchaseParams``:
+ *
+ * ```swift
+ * let params = PurchaseParams.Builder()
+ *                            .with(metadata: ["key": "value"])
+ *                            .with(promotionalOffer: promotionalOffer)
+ *                            .build()
+ *  Purchases.shared.purchase(package: package, params: params)
+ * ```
+ */
+@objc(RCPurchaseParams) public final class PurchaseParams: NSObject {
+
+    let promotionalOffer: PromotionalOffer?
```

Comment:
We already have [StoreProductDiscount](https://github.com/RevenueCat/purchases-ios/blob/main/Sources/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift) which wraps StoreKit's [SubcriptionOffer](https://developer.apple.com/documentation/storekit/product/subscriptionoffer). We will probably have to add for the `winBack` type to it when we add manual `winBackOffer` purchasing.

Unfortunately, the `SubscriptionOffer` object does not contain the signed data required to purchase a promo offer. That's why we have [a method](https://github.com/RevenueCat/purchases-ios/blob/main/Sources/Purchasing/Purchases/PurchasesType.swift#L664) to create a `PromotionalOffer` from a `StoreProductDiscount` and we need a separate `.with(promotionalOffer: promotionalOffer)` method that accepts it.

It would be great if we could have just a `.with(storeProductDiscount: storeProductDiscount)` method that could work for all types of offers tho.

- For intro offers, it would be applied automatically, so we wouldn't need to do anything with it.
- For promo offers, we would generate the signature and perform the purchase, or throw an error if the  signature is not valid or the customer is not eligible.
- For wingback offers, we would have already validated eligibility via `eligibleWinBackOfferIDs`, or otherwise the purchase method will just return an error.

Another option could be just specifying the priority order in which the offers will be applied, so e.g. if you include a winback offer and a promo offer, the winback will be applied and the promo offer will be ignored.

Reply:
Yeah... I think probably logging some error and prioritizing to one of the offers should work... As for @fire-at-will's suggestion, could it work if we just made it a protocol, and have both `PromotionalOffer`, `WinBackOffer` and anything else implement that protocol? Then we would accept either in the `with` function, but only one.